texk/ttfdump/docs/ttfdump.1
.TH TTFDUMP 1 "14 September 2025" "TeX Live" 
.SH NAME
ttfdump \- Dumping the contents of a TrueType Font file
.SH SYNOPSIS
.B ttfdump
.RB [ \-h ]
.RB [ \-t
.IR tablename ]
.RB [ \-g
.IR glyphnumber ]
.RB [ \-c
.IR collection ]
.RB [ \-o
.IR dumpfile ]
.RB [ \-i ]
.I ttfile
.SH DESCRIPTION
\fBttfdump\fP dumps the contents of a TrueType font file in ASCII form. 
A TrueType font file is consist of various tables. Those table are divided into
two categories - \fIRequired Tables\fP and \fIOptional Tables\fP. Mulitiple 
TrueType fonts can be merged to a TrueType Collection.
.SS Options
.TP
\fB-h\fP
Print a help message on standard output and exit.
.TP
\fB-t\fI tablename\fR
If this option is specified then dump the table \fI"tablename"\fP, otherwise 
dump every tables in \fIttfile\fP.
\fITablename\fP can be one of the following:
.RS
.TP
\fBRequired Tables\fR
.TP
\fIcmap\fP 
character to glyph mapping table
.TP
\fIglyf\fP
glyph data
.TP
\fIhead\fP
font header
.TP
\fIhhea\fP
horizontal header
.TP
\fIhmtx\fP
horizontal metrics
.TP
\fIloca\fP
index to location
.TP
\fImaxp\fP
maximum profile
.TP
\fIname\fP
naming table
.TP
\fIpost\fP
PostScript information
.TP
\fIOS/2\fP
OS/2 and windows specific metrics
.TP
\fBOptional Tables\fR
.TP
\fIcvt\fP
control value table
.TP
\fIfpgm\fP
font program; \fInot implemented yet.\fP
.TP
\fIgasp\fP
grid-fitting and scan conversion procedure (grayscale)
.TP
\fIhdmx\fP
horizontal device metrics
.TP
\fIkern\fP
kerning table
.TP
\fILTSH\fP
Linear threshold table
.TP
\fIprep\fP
CVT program; \fInot implemented yet\fP
.TP
\fIPCLT\fP
PCL5 table
.TP
\fIVDMX\fP
Vertical Device Metrics table
.TP
\fIvhea\fP
vertical metrics header
.TP
\fIvmtx\fP
vertical metrics
.TP
\fBAdditional OpenType Tables\fR
.TP
\fIGPOS\fP
glyph positioning table
.TP
\fIGSUB\fP
glyph substitution table
.RE
.TP
\fB-g\fI glyphnumber\fR
If this option is specified then dump the glyph with index \fIglyphnumber\fP, 
otherwise dump all glyphs in \fIttfile\fP.
.TP
\fB-c\fI collection\fR
Select a TrueType font in a TrueType Collection.
.TP
\fB-o\fI dumpfile\fR
Output of the dump. If this option is not specified, dump to stdout.
.TP
\fB-i\fI ttfile\fR
TrueType font file to be dumped.
.RE
.SH EXAMPLE
To dump all information in the TrueType font file times.ttf:
.RS
ttfdump times.ttf
.RE

To dump the required table \fIcmap\fP in the file:
.RS
ttfdump -t cmap times.ttf
.RE

To dump a single glyph with index 50:
.RS
ttfdump -g 50 times.ttf
.RE

.SH SEE ALSO
TrueType 1.0 Font Files, \fI Technical Specification Revision 1.66 November 1995\fR
.SH NOTE
Instruction disassembler in not implemented.
.SH AUTHOR
Li-Da Lho
.PP
Currently maintained as part of TeX Live; email bug reports to
tex-k@tug.org..


ChangeLog
2025-09-07  Karl Berry  <karl@freefriends.org>

	* Makefile.am (install_target): new variable, install-strip by default.
	(world): use it.

2025-07-22  Karl Berry  <karl@freefriends.org>

	* reautoconf: mention that autoconf-only packages need to be
	explicitly added; would be better to read kpse-pkgs.m4.
	Also reautoconf tree for kpse-setup.m4 changes.

2025-06-26  Karl Berry  <karl@freefriends.org>

	* automake 1.18.1.

2025-04-30  Karl Berry  <karl@freefriends.org>

	* Build: incorporate original CFLAGS, CXXFLAGS, OBJCXXFLAGS values.
	https://tug.org/pipermail/tlbuild/2025q2/005771.html

2025-04-29  Karl Berry  <karl@freefriends.org>

	* Build: use TL_CFLAGS, TL_CXXFLAGS, etc., for TL_BUILD_ENV
	if -g is not given.

2025-03-08  Karl Berry  <karl@tug.org>

	* version.ac: 2026/dev.

2025-03-07  Karl Berry  <karl@tug.org>

	* TL'25 release.

2025-01-12  Karl Berry  <karl@tug.org>

	* Build: also support --debug-more for -g -Og -ggdb3,
	per Luigi, as described on https://tug.org/texlive/build.html.

2024-07-12  Karl Berry  <karl@freefriends.org>

	* reautoconf with automake 1.17.

2024-04-29  Karl Berry  <karl@freefriends.org>

	* Makefile.am (distdir): override default dist generation,
	since we don't try to keep make dist working any more.
	https://tug.org/pipermail/tlbuild/2024q2/005547.html

2023-12-31  Karl Berry  <karl@freefriends.org>

	* reautoconf: add -f / --force option to pass to autoreconf.

2023-12-26  Karl Berry  <karl@freefriends.org>

	* reautoconf: don't bother with environment variables;
	because the autotools have many subprograms, setting PATH
	is what makes sense.

2023-12-23  Karl Berry  <karl@tug.org>

	* reautoconf with autoconf-2.72.

2023-12-21  Karl Berry  <karl@freefriends.org>

	* reautoconf (do_it): abort the run if any command fails.

2023-02-25  Karl Berry  <karl@tug.org>

	* Build: show full uname -a.

2023-02-09  Karl Berry  <karl@freefriends.org>

	* version.ac: 2023, no more /dev.

2022-03-21  Karl Berry  <karl@freefriends.org>

	* tardate.ac (tex_live_tardate): 2022-03-21 for TL'22.

2022-01-18  Karl Berry  <karl@freefriends.org>

	* version.ac: 2022, no more /dev.

2021-10-04  Karl Berry  <karl@freefriends.org>

	* reautoconf with automake-1.16.5.

2021-06-27  Karl Berry  <karl@freefriends.org>

	* reautoconf: pass -Wno-obsolete by default, since Autoconf 2.70
	and later obsoletes many common macros. See m4/README for more.
	Also standardize shell usage, etc., but no other functional changes.

2021-04-17  Karl Berry  <karl@tug.org>

	* version.ac: now 2022/dev since TL'21 is released.

2021-03-23  Karl Berry  <karl@tug.org>

	* tardate.ac (tex_live_tardate): 2020-03-23 for TL'21.

2021-02-17  Karl Berry  <karl@tug.org>

	* version.ac: 2021, no more /dev.

2020-05-28  Karl Berry  <karl@freefriends.org>

	* Makefile.am (world): do not descend into $(texlinks_dir) if it
	doesn't exist. Report from Andreas Scherer,
	https://tug.org/pipermail/tlbuild/2020q1/004614.html.

2020-05-14  Karl Berry  <karl@freefriends.org>

	* reautoconf for removal of poppler support.

2020-04-10  Karl Berry  <karl@tug.org>

	* version.ac: 2020/dev.

2020-03-27  Karl Berry  <karl@tug.org>

	* tardate.ac (tex_live_tardate): 2020-03-27 for TL'20.

2020-03-06  Karl Berry  <karl@tug.org>

	* .travis.yml: attempt to reenable musl build on github.

2020-01-31  Karl Berry  <karl@tug.org>

	* Makefile.am (texlinks_dir): now texk/texlive/tl_support
	(instead of tl_scripts).

2020-01-16  Karl Berry  <karl@freefriends.org>

	* version.ac (tex_live_version): remove /dev, preparing for TL'20.

2019-06-25  Karl Berry  <karl@freefriends.org>

	* Build: don't call buildinfo from here, the output when the CC
	etc. envvars are not set is too misleading. Just write UNAME.

2019-06-24  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh (compiler_version): new shell function;
	 ignore useless first "Configured:" line for gcc->cc on Macs.

2019-06-23  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh: output the compiler/other names
	that we pass from the buildinfo target.
	* Makefile.am (buildinfo): pass OBJCXX, not OBJC.
	* Build: more maintainable exports of TL_*.

2019-06-22  Karl Berry  <karl@freefriends.org>

	* doc/buildinfo.sh: rename from build-info.sh;
	forget /etc/issue;
	allow option to omit config.status output;
	do not dump environment;
	run lsb_release -a. We'll see ...
	* Build: use buildinfo.sh instead of inline commands (mostly).
	* Makefile.am (buildinfo): make dependency of all-recursive
	so that it runs first; rename from build-info; remove from world.

2019-06-06  Karl Berry  <karl@tug.org>

	* Makefile.am (world): make build-info.
	(build-info): new target to invoke doc/build-info.sh.
	(build_info_target): new variable.
	* Build: export TL_* variables so build-info.sh can report them.
	* doc/build-info.sh: new script to report compiler versions,
	dump environment, config.status, etc.

2019-05-03  Karl Berry  <karl@tug.org>

	* version.ac: 2020/dev.

2019-04-07  Karl Berry  <karl@freefriends.org>

	* tardate.ac (tex_live_tardate): let's hope for 2019-04-07.

2018-10-02  Karl Berry  <karl@tug.org>

	* Build: TL_MAKE_VERBOSE=VERBOSE=1 by default,
	so that test-suite.log will be shown on failure.
	https://lists.gnu.org/archive/html/automake/2018-09/msg00015.html

2018-04-28  Karl Berry  <karl@tug.org>

	* version.ac: 2019/dev.

2018-04-10  Karl Berry  <karl@tug.org>

	* README.*: update from new tlbuild.

2017-11-15  Karl Berry  <karl@tug.org>

	* *.test: use #!/bin/sh -vx. Add 

2017-03-21  Karl Berry  <karl@freefriends.org>

	* everywhere: use EXTRA_DIST=TLpatches instead of <pkgname>-PATCHES.

2017-03-18  Karl Berry  <karl@tug.org>

	* version.ac (tex_live_version): 2017/dev => 2017 for TL17.

2016-11-16  Karl Berry  <karl@tug.org>

	* Makefile.am (install-exec-hook): warn that the
	result of make install is not in itself usable.

2016-06-30  Karl Berry  <karl@tug.org>

	* version.ac (tex_live_version): 2016 => 2017/dev.

texk/ttfdump/src/ttfdump.c
/* ttfdump - dump contents of a TrueType font file
   $Id$
   Originally written by Li-Da Lho, now maintained as part of TeX Live.
   Released under the GNU GPL (any version).
*/

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#if defined(KPATHSEA)
/* including <kpathsea/kpathsea.h> would yield conflicting typedefs */
#include <kpathsea/lib.h>
#include <kpathsea/tex-file.h>
#include <kpathsea/c-fopen.h>
#include <kpathsea/progname.h>
#include <kpathsea/c-unistd.h>
# if defined (WIN32) 
#include <kpathsea/getopt.h>
# else
#include <unistd.h>
# endif
#elif defined(WIN32)
#include <getopt.h>
#else
#include <unistd.h>
#endif

#include <string.h>
#include <stdio.h>
#include <stdlib.h>

#include "ttf.h"
#include "ttfutil.h"
#include "ttc.h"

/* $Id$  */

#define MAXLEN 256
#define ALL_GLYF -1
#define NO_GLYF -2

static char ttfname[MAXLEN], dumpname[MAXLEN];
static TTFontPtr font;

static void print_ttc(TTCHeaderPtr ttc, FILE *out);
static void print_prologue(FILE *out);
static void print_offset(FILE *out);
static void print_dir(FILE *out);
static void print_all_tables(FILE *dp_file);
static void print_table(FILE *out, char *tablename);

static void print_cmap(FILE *out);
static void print_glyf(FILE *out, ULONG index);
static void print_all_glyfs(FILE *out);
static void print_head(FILE *out);
static void print_hhea(FILE *out);
static void print_hmtx(FILE *out);
static void print_loca(FILE *out);
static void print_maxp(FILE *out);
static void print_name(FILE *out);
static void print_post(FILE *out);
static void print_os2(FILE *out);

static void print_cvt(FILE *out);
static void print_fpgm(FILE *out);
static void print_gasp(FILE *out);
static void print_hdmx(FILE *out);
static void print_kern(FILE *out);
static void print_ltsh(FILE *out);
static void print_pclt(FILE *out);
static void print_prep(FILE *out);
static void print_vdmx(FILE *out);
static void print_vhea(FILE *out);
static void print_vmtx(FILE *out);

static void print_gpos(FILE *out);
static void print_gsub(FILE *out);

static void usage (void);


int
main(int argc, char *argv[])
{
  FILE *dp_file;
  TTCHeaderPtr ttc;

  int c;
  extern int optind;
  extern char *optarg;

  char *tablename = NULL;


  /* print all glyphs by default */
  int glyphnum = ALL_GLYF, collection = 0;

#ifdef KPATHSEA
  kpse_set_program_name(argv[0], "ttfdump");
#endif

  if (argc < 2)
  {
    usage();
    exit(EXIT_FAILURE);
  }

  while ((c = getopt(argc, argv, "t:g:c:i:o:h")) != EOF)
  {
    switch (c)
    {
    case 't':
      tablename = xstrdup(optarg);
      break;
    case 'g':
      if (strcmp(optarg, "x") == 0)
        glyphnum = NO_GLYF;
      else
        glyphnum = atoi(optarg);
      break;
    case 'c':
      collection = atoi(optarg);
      break;
    case 'i':
      if(strlen(optarg) >= MAXLEN) {
        fprintf(stderr, "Too long input file name\n");
        exit(EXIT_FAILURE);
      }
      strcpy(ttfname, optarg);
      break;
    case 'o':
      if(strlen(optarg) >= MAXLEN) {
        fprintf(stderr, "Too long output file name\n");
        exit(EXIT_FAILURE);
      }
      strcpy(dumpname, optarg);
      break;
    case 'h':
    case '?':
      usage();
      exit(EXIT_FAILURE);
      break;
    }
  }

  /* processing ttf file if -i flag is not given */
  if (*ttfname == 0 && optind < argc)
  {
    if(strlen(argv[optind]) >= MAXLEN) {
      fprintf(stderr, "Too long input file name\n");
      exit(EXIT_FAILURE);
    }
    strcpy(ttfname, argv[optind]);
  }

  /* processing dumping file if -o flag is not given */
  if (*dumpname == 0)
  {
    if (optind + 1 < argc)
    {
      /* no -o flag but dumping filename specified */
      strcpy(dumpname, argv[optind + 1]);
    }
    else
      strcpy(dumpname, "-");
  }
  if (!strcmp(dumpname, "-"))
  {
#ifdef WIN32
    _setmode(fileno(stdout), _O_BINARY);
#endif
    dp_file = stdout;
  }
  else if ((dp_file = fopen(dumpname, "wb")) == NULL)
  {
    fprintf(stderr, "Can't open dumping file\n");
    exit(EXIT_FAILURE);
  }

  /* Try to load as a ttc file. */
  ttc = ttfLoadTTCHeader(ttfname);

  if (ttc != NULL)
  {
    if (collection < ttc->DirCount)
      font = ttc->font + collection;
    else
      fprintf (stderr, "TrueType collection number too large\n"
               "should between 0 and %d\n", ttc->DirCount - 1);
  }
  else
    /* If ttc failed, try to load as a ttf. */
    font = ttfInitFont(ttfname);

  if (font == NULL)
    exit(EXIT_FAILURE);

  print_prologue(dp_file);
  if (ttc != NULL)
  {
    print_ttc(ttc, dp_file);
  }
  print_offset(dp_file);
  print_dir(dp_file);

  if (font->loca == NULL)
    glyphnum = NO_GLYF;

  if (tablename == NULL)
  {
    /* no table specified */
    print_all_tables(dp_file);
    if (glyphnum == ALL_GLYF)
      print_all_glyfs(dp_file);
    else if (glyphnum == NO_GLYF)
      ;
    else
      print_glyf(dp_file, glyphnum);
  }
  else
  {
    /* table specified */
    print_table(dp_file, tablename);
    if (!strcmp(tablename, "glyf"))
    {
      if (glyphnum == ALL_GLYF)
        print_all_glyfs(dp_file);
      else if (glyphnum == NO_GLYF)
        ;
      else
        print_glyf(dp_file, glyphnum);
    }
  }

  if (ttc != NULL)
  {
    ttfFreeTTCFont(ttc);
  }
  else
  {
    ttfFreeFont(font);
  }
  return 0;
}


static void
print_table(FILE *out, char *tablename)
{
  if (!strcmp(tablename, "cmap"))
    print_cmap(out);
  if (!strcmp(tablename, "head"))
    print_head(out);
  if (!strcmp(tablename, "hhea"))
    print_hhea(out);
  if (!strcmp(tablename, "hmtx"))
    print_hmtx(out);
  if (!strcmp(tablename, "loca"))
    print_loca(out);
  if (!strcmp(tablename, "maxp"))
    print_maxp(out);
  if (!strcmp(tablename, "name"))
    print_name(out);
  if (!strcmp(tablename, "post"))
    print_post(out);
  if (!strcmp(tablename, "OS/2"))
    print_os2(out);
  if (!strcmp(tablename, "cvt"))
    print_cvt(out);
  if (!strcmp(tablename, "fpgm"))
    print_fpgm(out);
  if (!strcmp(tablename, "gasp"))
    print_gasp(out);
  if (!strcmp(tablename, "hdmx"))
    print_hdmx(out);
  if (!strcmp(tablename, "kern"))
    print_kern(out);
  if (!strcmp(tablename, "LTSH"))
    print_ltsh(out);
  if (!strcmp(tablename, "prep"))
    print_prep(out);
  if (!strcmp(tablename, "VDMX"))
    print_vdmx(out);
  if (!strcmp(tablename, "vhea"))
    print_vhea(out);
  if (!strcmp(tablename, "vmtx"))
    print_vmtx(out);
  if (!strcmp(tablename, "GPOS"))
    print_gpos(out);
  if (!strcmp(tablename, "GSUB"))
    print_gsub(out);
}


static void
print_all_tables(FILE *dp_file)
{
  print_cmap(dp_file);
  print_head(dp_file);
  print_hhea(dp_file);
  print_hmtx(dp_file);
  print_loca(dp_file);
  print_maxp(dp_file);
  print_name(dp_file);
  print_post(dp_file);
  print_os2(dp_file);

  /*optional tables */
  print_cvt(dp_file);
  print_gasp(dp_file);
  print_hdmx(dp_file);
  print_kern(dp_file);
  print_ltsh(dp_file);
  print_pclt(dp_file);
  print_vdmx(dp_file);
  print_vhea(dp_file);
  print_vmtx(dp_file);

  /* Additional OpenType Tables */
  print_gpos(dp_file);
  print_gsub(dp_file);
}


static void
print_ttc(TTCHeaderPtr ttc, FILE *out)
{
  int i, b[2];

  FixedSplit(ttc->version, b);

  fprintf(out, "TrueType Collection Header\n");
  fprintf(out, "--------------------------\n");
  fprintf(out, "TTC version: \t\t %d.%d\n", b[1], b[0]);
  fprintf(out, "Number of fonts:\t %d\n", ttc->DirCount);
  for (i = 0; i < ttc->DirCount; i++)
    fprintf(out, "Offset of Directory #%d:\t %d\n", i, ttc->offset[i]);
  fprintf(out, "\n");
}


static void
print_prologue(FILE *out)
{
  fprintf(out, "True Type Font File Dumper: v 0.5.5\n");
  fprintf(out, "Copyright 1996-1998 ollie@ms1.hinet.net\n");
  fprintf(out, "Dumping File:%s\n\n\n", font->ttfname);
}


static void
print_offset(FILE *out)
{
  int b[2];

  FixedSplit(font->version, b);

  fprintf(out, "Offset Table\n");
  fprintf(out, "------------\n");
  fprintf(out, "\t sfnt version: ");
  if (b[1] & 0xff00)
    fprintf(out, "'%s'\n", TagToStr (font->version));
  else
    fprintf(out, "\t\t %d.%d\n", b[1], b[0]);
  fprintf(out, "\t number of tables: %d\n", font->numTables);
}


static void
print_dir(FILE *out)
{
  int i;


  for (i = 0; i < font->numTables; i++)
  {
    fprintf(out, "  %2d. ", i);
    ttfPrintTableDir(out, font->dir + i);
  }
  fprintf(out, "\n");
}

static void
print_cmap(FILE *out)
{
  ttfPrintCMAP(out, font->cmap);
  fprintf(out, "\n");
}

static void
print_glyf(FILE *out, ULONG index)
{
  GLYFPtr glyf;


  glyf = ttfLoadGlyphIndex(font, index);
  fprintf(out, "Glyph %6d.\n", index);
  ttfPrintGLYF(out, glyf);
}


static void
print_all_glyfs(FILE *out)
{
  int i;
  GLYFPtr glyf;


  for (i = 0; i < font->maxp->numGlyphs; i++)
  {
    glyf = ttfLoadGlyphIndex(font, i);
    fprintf(out, "Glyph %6d.\n", i);
    ttfPrintGLYF(out, glyf);
  }
}


static void
print_head(FILE *out)
{
  ttfPrintHEAD(out, font->head);
  fprintf(out, "\n");
}


static void
print_hhea(FILE *out)
{
  ttfPrintHHEA(out, font->hhea);
  fprintf(out, "\n");
}


static void
print_hmtx(FILE *out)
{
  /* because of interdependencies between tables
   * we have to init hhea and maxp before hmtx */
  ttfPrintHMTX(out, font->hmtx);
  fprintf(out, "\n");
}


static void
print_loca(FILE *out)
{
  /* because of interdependencies between tables
   * we have to init head and maxp before loca */
  ttfPrintLOCA(out, font->loca);
  fprintf(out, "\n");
}


static void
print_maxp(FILE *out)
{
  ttfPrintMAXP(out, font->maxp);
  fprintf(out, "\n");
}


static void
print_name(FILE *out)
{
  ttfPrintNAME(out, font->name);
  fprintf(out, "\n");
}


static void
print_post(FILE *out)
{
  ttfPrintPOST(out, font->post);
}

static void
print_os2(FILE *out)
{
  ttfPrintOS2(out, font->os2);
}


static void
print_cvt(FILE *out)
{
  if (font->cvt != NULL)
    ttfPrintCVT(out, font->cvt, font->cvtLength);
}


static void
print_fpgm(FILE *out)
{
  if (font->fpgm != NULL)
    ttfPrintFPGM(out, font->fpgm, font->fpgmLength);
}


static void
print_gasp(FILE *out)
{
  if (font->gasp != NULL)
    ttfPrintGASP(out, font->gasp);
}


static void
print_hdmx(FILE *out)
{
  if (font->hdmx != NULL)
    ttfPrintHDMX(out, font->hdmx);
}


static void
print_kern(FILE *out)
{
  if (font->kern != NULL)
    ttfPrintKERN(out, font->kern);
}


static void
print_ltsh(FILE *out)
{
  if (font->ltsh != NULL)
    ttfPrintLTSH(out, font->ltsh);
}


static void
print_pclt(FILE *out)
{
  if (font->pclt != NULL)
    ttfPrintPCLT(out, font->pclt);
}


static void
print_prep(FILE *out)
{
  if (font->prep != NULL)
    ttfPrintPREP(out, font->prep, font->prepLength);
}


static void
print_vdmx(FILE *out)
{
  if (font->vdmx != NULL)
    ttfPrintVDMX(out, font->vdmx);
}


static void
print_vhea(FILE *out)
{
  if (font->vhea != NULL)
    ttfPrintVHEA(out, font->vhea);
}


static void
print_vmtx(FILE *out)
{
  if (font->vmtx != NULL)
    ttfPrintVMTX(out, font->vmtx);
}


static void
print_gpos(FILE *out)
{
  if (font->gpos != NULL)
    ttfPrintGPOS(out, font->gpos);
}


static void
print_gsub(FILE *out)
{
  if (font->gsub != NULL)
    ttfPrintGSUB(out, font->gsub);
}


static void
usage(void)
{
  fprintf(stderr, "ttfdump %s (%s)\n", VERSION, TL_VERSION);
  fprintf(stderr, "Usage: ttfdump ttfname dumpname [options]\n");
  fprintf(stderr, "\nEmail bug reports to %s.\n", PACKAGE_BUGREPORT);
}


/* end of ttfdump.c */
